File: 9.Minimum-Spanning-Tree_Shortest_Path/jiyeon/3.network-delay-time.py
# https://leetcode.com/problems/network-delay-time/

import collections
import heapq
from typing import List

# FIXME!


class Solution:
    def networkDelayTime(self, times: List[List[int]], end_point: int, start_point: int) -> int:
        dicts = collections.defaultdict(dict)
        for curr_node, next_node, distance in times:
            dicts[curr_node][next_node] = distance

        # Min-heap to store minimum weight edge at top.
        heap = [(0, start_point)]

        # Track nodes which are included in MST.
        visited = [False] * end_point

        mst_cnt = len(times)

        while heap:
            distance, idx = heapq.heappop(heap)
            if idx == end_point:
                return distance

            if visited[idx]:
                continue
            visited[idx] = True
            for curr_node, dict_distance in dicts[idx].items():
                heapq.heappush(
                    heap, (distance + dict_distance, curr_node))
            mst_cnt - 1
        return -1


print(Solution().networkDelayTime([[2, 1, 1], [2, 3, 1], [3, 4, 1]], 4, 2))
# print(Solution().networkDelayTime([[1, 2, 1]], 2, 2))
# print(Solution().networkDelayTime([[1, 2, 1], [2, 1, 3]], 2, 2))
# print(Solution().networkDelayTime([[1,2,1],[2,3,2],[1,3,1]], 3, 2))

File: 9.Minimum-Spanning-Tree_Shortest_Path/동규/3.network-delay-time.js
// hdelayTimeps://leetcode.com/problems/network-delay-time/

var networkDelayTime = function (times, N, K) {
  const delayTime = new Array(N + 1).fill(Infinity);
  delayTime[0] = 0;
  delayTime[K] = 0;

  let flag = true;

  while (flag) {
    flag = false;
    times.forEach(([u, v, w]) => {
      if (delayTime[u] !== Infinity && delayTime[v] > delayTime[u] + w) {
        delayTime[v] = delayTime[u] + w;
        flag = true;
        // console.log("delayTime[u]", delayTime[u]);
        // console.log("delayTime[v]", delayTime[v]);
        console.log("delayTime[v]", delayTime[v]);
      }
    });
  }

  const res = Math.max(...delayTime);

  return res === Infinity ? -1 : res;
};

networkDelayTime(
  [
    [2, 1, 1],
    [2, 3, 1],
    [3, 4, 1],
  ],
  4,
  2
);
